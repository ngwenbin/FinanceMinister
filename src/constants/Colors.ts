const primary = {
  100: "#CCE1FE",
  200: "#9AC1FE",
  300: "#689EFD",
  500: "#054EFA",
  700: "#022CB3",
  800: "#011F90",
};

const secondary = {
  200: "#FFECB1",
  500: "#FFBC3C",
  900: "#7A440B",
};

const gray = {
  50: "#F9FAFB",
  100: "#F3F4F6",
  200: "#E5E7EB",
  300: "#D1D5DB",
  400: "#9CA3AF",
  500: "#6B7280",
  600: "#4B5563",
  700: "#374151",
  800: "#1F2937",
  900: "#111827",
};

const error = {
  100: "#FFE7D3",
  200: "#FFC9A6",
  500: "#FF4823",
  700: "#B71411",
  900: "#7A0616",
};

const warning = {
  100: "#FFF8CF",
  200: "#FFF09F",
  500: "#FFCB11",
  700: "#B78908",
  900: "#7A5503",
};

const success = {
  100: "#EEFCD1",
  200: "#DAFAA5",
  500: "#72D11F",
  700: "#40960F",
  900: "#1D6405",
};

const info = {
  100: "#DDEFFE",
  200: "#BDDEFE",
  500: "#5997F9",
  700: "#2C56B3",
  900: "#112877",
};

export const Colors = {
  primary,
  secondary,
  gray,
  error,
  warning,
  success,
  info,
  white: "#FFFFFF",
  black: "#000000",
};
