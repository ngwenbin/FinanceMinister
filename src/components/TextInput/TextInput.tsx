import {
  StyleSheet,
  View,
  TextStyle,
  TextInput as DefaultTextInput,
  TextInputProps as DefaultTextInputProps,
} from "react-native";
import { useState } from "react";

import { Text } from "@components/Text";
import { TextVariantsData } from "@constants/Fonts";
import { Colors } from "@constants/Colors";
import { getFont } from "@utils";
import { testIds } from "@utils/tests/testIds";
import { ColorValue } from "react-native";

interface TextInputProps extends Omit<DefaultTextInputProps, "multiline"> {
  label: string;
  labelStyle?: TextStyle;
  required?: boolean;
  subtext?: string; // use it for error message
  subtextStyle?: TextStyle;
  globalColorOverwrite?: ColorValue;
}

const TextInput = ({
  label,
  labelStyle,
  required,
  subtext,
  subtextStyle,
  globalColorOverwrite,
  ...props
}: TextInputProps) => {
  const { variation, size } = TextVariantsData["body-small"];

  const [isFocused, setIsFocused] = useState<boolean>(false);
  return (
    <View style={styles.container}>
      <Text
        testID={testIds.sharedComponents.textInput.label}
        style={[styles.label, labelStyle]}
        variant="body-small"
      >
        {label}
        {required ? (
          <Text
            color={Colors.error[500]}
            variant="body-small"
            testID={testIds.sharedComponents.textInput.labelAdornment}
          >
            *
          </Text>
        ) : null}
      </Text>
      <DefaultTextInput
        testID={testIds.sharedComponents.textInput.input}
        style={[
          getFont(variation, size),
          { borderColor: isFocused ? Colors.primary[500] : Colors.gray[300] },
          styles.textInputContainer,
          props.style,
          ...(globalColorOverwrite
            ? [
                {
                  borderColor: globalColorOverwrite,
                  color: globalColorOverwrite,
                },
              ]
            : []),
        ]}
        onFocus={() => setIsFocused(true)}
        onBlur={() => setIsFocused(false)}
        {...props}
        multiline={false}
        cursorColor={Colors.gray[700]}
      />
      {subtext ? (
        <Text
          testID={testIds.sharedComponents.textInput.subtext}
          style={[
            getFont("light", "xs"),
            styles.subtext,
            subtextStyle,
            ...(globalColorOverwrite
              ? [
                  {
                    color: globalColorOverwrite,
                  },
                ]
              : []),
          ]}
        >
          {subtext}
        </Text>
      ) : null}
    </View>
  );
};

export default TextInput;

const styles = StyleSheet.create({
  container: {
    maxWidth: 320,
    width: "100%",
    margin: 4,
  },
  label: {
    color: Colors.gray[900],
  },
  textInputContainer: {
    paddingHorizontal: 16,
    paddingVertical: 10,
    borderWidth: 1,
    borderRadius: 8,
    marginTop: 6,
    marginBottom: 2,
  },
  subtext: {
    color: Colors.gray[500],
  },
});
