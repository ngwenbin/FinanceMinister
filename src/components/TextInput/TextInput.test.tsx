import { render, screen } from "@testing-library/react-native";
import "@testing-library/jest-native/extend-expect";

import TextInput from "./TextInput";
import { testIds } from "@utils/tests/testIds";
import { Colors } from "@constants/Colors";

const textInputLabel = "This is a text label";
const textSubtext = "This is a subtext";
const labelTestId = testIds.sharedComponents.textInput.label;
const labelAdornmentTestId = testIds.sharedComponents.textInput.labelAdornment;
const textinputfieldTestId = testIds.sharedComponents.textInput.input;
const subtextTestId = testIds.sharedComponents.textInput.subtext;

describe("<TextInput>", () => {
  describe("as default", () => {
    test("Renders successfully", () => {
      const activeText = render(<TextInput label={textInputLabel} />);
      expect(activeText).toBeDefined();
    });

    test("Renders correctly", () => {
      const activeText = render(
        <TextInput label={textInputLabel} subtext={textSubtext} />
      );
      expect(activeText.toJSON()).toMatchSnapshot();
    });

    test("Outputs the correct text", () => {
      render(<TextInput label={textInputLabel} subtext={textSubtext} />);
      expect(screen.getByTestId(labelTestId)).toHaveTextContent(textInputLabel);
      expect(screen.getByTestId(subtextTestId)).toHaveTextContent(textSubtext);
    });

    test("Outputs the correct subtext style", () => {
      render(
        <TextInput
          label={textInputLabel}
          subtext={textSubtext}
          subtextStyle={{ color: Colors.primary[500] }}
        />
      );
      expect(screen.getByTestId(subtextTestId)).toHaveStyle({
        color: Colors.primary[500],
      });
    });

    test("Outputs the correct global color overwrite", () => {
      render(
        <TextInput
          label={textInputLabel}
          subtext={textSubtext}
          subtextStyle={{ color: Colors.primary[500] }}
          globalColorOverwrite={Colors.error[500]}
        />
      );
      expect(screen.getByTestId(textinputfieldTestId)).toHaveStyle({
        color: Colors.error[500],
        borderColor: Colors.error[500],
      });
      expect(screen.getByTestId(subtextTestId)).toHaveStyle({
        color: Colors.error[500],
      });
    });

    test("Shows required adornment if enabled", () => {
      render(
        <TextInput label={textInputLabel} subtext={textSubtext} required />
      );
      expect(screen.queryByTestId(labelAdornmentTestId)).toBeDefined();
    });

    test("Hides required adornment if disabled", () => {
      render(<TextInput label={textInputLabel} subtext={textSubtext} />);
      expect(screen.queryByTestId(labelAdornmentTestId)).toBeNull();
    });
  });
});
