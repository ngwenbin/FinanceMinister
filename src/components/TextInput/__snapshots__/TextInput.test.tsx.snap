// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput> as default Renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 4,
      "maxWidth": 320,
      "width": "100%",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#111827",
        },
        Object {
          "fontFamily": "Inter_400Regular",
          "fontSize": 14,
        },
        Array [
          Object {
            "color": "#111827",
          },
          undefined,
        ],
      ]
    }
    testID="textinput-0"
  >
    This is a text label
  </Text>
  <TextInput
    cursorColor="#374151"
    multiline={false}
    onBlur={[Function]}
    onFocus={[Function]}
    style={
      Array [
        Object {
          "fontFamily": "Inter_400Regular",
          "fontSize": 14,
        },
        Object {
          "borderColor": "#D1D5DB",
        },
        Object {
          "borderRadius": 8,
          "borderWidth": 1,
          "marginBottom": 2,
          "marginTop": 6,
          "paddingHorizontal": 16,
          "paddingVertical": 10,
        },
        undefined,
      ]
    }
    testID="textinput-2"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#111827",
        },
        Object {
          "fontFamily": "Inter_400Regular",
          "fontSize": 12,
        },
        Array [
          Object {
            "fontFamily": "Inter_300Light",
            "fontSize": 12,
          },
          Object {
            "color": "#6B7280",
          },
          undefined,
        ],
      ]
    }
    testID="textinput-3"
  >
    This is a subtext
  </Text>
</View>
`;
